src/hotspot/share/gc/shenandoah/heuristics/shenandoahGenerationalHeuristics.cpp

```diff
@@ -216,6 +218,19 @@ void ShenandoahGenerationalHeuristics::choose_collection_set(ShenandoahCollectio
                        byte_size_in_proper_unit(promote_evac_bytes), proper_unit_for_byte_size(promote_evac_bytes),
                        byte_size_in_proper_unit(old_evac_bytes), proper_unit_for_byte_size(old_evac_bytes),
                        byte_size_in_proper_unit(total_evac_bytes), proper_unit_for_byte_size(total_evac_bytes));
+
+    ShenandoahEvacInfo evacInfo;
+    evacInfo.set_collection_set_regions(collection_set->count());
+    evacInfo.set_collection_set_used_before(collection_set->used());
+    evacInfo.set_collection_set_used_after(collection_set->live());
+    evacInfo.set_collected_old(old_evac_bytes);
+    evacInfo.set_collected_promoted(promote_evac_bytes);
+    evacInfo.set_collected_young(young_evac_bytes);
+    evacInfo.set_regions_immediate(immediate_regions);
+    evacInfo.set_immediate_size(immediate_garbage);
+    evacInfo.set_regions_freed(free_regions);
+
+    ShenandoahTracer().report_evacuation_info(&evacInfo);
```

Comment:
After we chatted, I thought of one other detail that would be nice to gather in this evacInfo event: regions to be promoted in place.  Associated with this detail, would be nice to know:
1. how many regions
2. how much garbage is known to exist within these regions
3. how much free memory is known to exist within these regions
Same for humongous regions promoted in place.

See humongous_regions_promoted, regular_regions_promoted_in_place, regular_regions_promoted_usage.   You might need to add another bookkeeping variable to count regular_regions_promoted_live, tracking regular_regions_promoted_usage.

Reply:
for humongous_regions_promoted, less important to distinguish used and free, because that memory is "off limits" to allocators.

Reply:
Looks good.